workers/unity/Assets/Playground/Scripts/Worker/AbstractMobileClientWorker.cs

```diff
@@ -0,0 +1,81 @@
+﻿using Improbable.Gdk.Core;
+using Improbable.Gdk.Mobile;
+using Playground;
+using UnityEngine;
+using UnityEngine.UI;
+
+public abstract class AbstractMobileClientWorker : MobileWorkerConnector
+{
+    public GameObject ConnectionPanel;
+    public GameObject Level;
+
+    private GameObject levelInstance;
+    private bool connected;
+    private InputField input;
+    private Button button;
+    private Text errorMessage;
+
+    public void Awake()
+    {
+        input = ConnectionPanel.transform.Find("ConnectInput").GetComponent<InputField>();
+        button = ConnectionPanel.transform.Find("ConnectButton").GetComponent<Button>();
+        errorMessage = ConnectionPanel.transform.Find("ConnectionError").GetComponent<Text>();
+
+        input.text = PlayerPrefs.GetString("cachedIp");
+    }
+
+    public void Start()
+    {
+        button.onClick.AddListener(Connect);
+    }
+
+    public async void Connect()
+    {
+        errorMessage.text = "";
+        await Connect(WorkerUtils.UnityClient, new ForwardingDispatcher()).ConfigureAwait(false);
+    }
+
+    protected override void HandleWorkerConnectionEstablished()
+    {
+        WorkerUtils.AddClientSystems(Worker.World);
+        if (Level == null)
+        {
+            return;
+        }
+
+        levelInstance = Instantiate(Level, transform);
+        levelInstance.transform.SetParent(null);
+
+        connected = true;
+        ConnectionPanel.SetActive(false);
+
+        PlayerPrefs.SetString("cachedIp", input.text);
+        PlayerPrefs.Save();
+    }
+
+    protected override void HandleWorkerConnectionFailure()
+    {
+        errorMessage.text = "Connection failure. Please check the IP address";
+    }
+
+    protected string GetIpFromField()
+    {
+        return input.text;
+    }
+
+    public override void Dispose()
+    {
+        if (levelInstance != null)
+        {
+            Destroy(levelInstance);
+        }
+
+        if (connected)
+        {
+            Destroy(this);
```

Comment:
this looks like we call destroy to early.

remove all of that except for the levelInstance part and just call `base.Dispose();`, see https://github.com/spatialos/gdk-for-unity/blob/a8e7bb33e8042062bae3cbe9c16cfea5472fd788/workers/unity/Assets/Playground/Scripts/Worker/ClientWorkerConnector.cs how it should look like

Reply:
Destroy is a deferred call. A gameobject doesn't get destroyed until the end of the tick.

Reply:
but still think we should call the base for this unless there is a reason for doing it like this

Reply:
The reason for doing it like this is allowing connection retry with the same GameObject i.e. to allow fixing a wrong IP, for example.

Reply:
Re-read this. Very against not calling base.Dispose(). EWe could easily change the base behaviour to do more and not know to change this. Why not make the thing that gets and IP something else and have that create a gamobject whenever it tries to make a connection?

Reply:
yeah, that's what should happen. create the worker gameobject only when necessary. reusing the current one could potentially cause issues

Reply:
+1, in a dispose if you don't call base.dispose if the base changes in the future it will be the result of very funky bugs that may be a pain to diagnose, if we fix the reason that makes it inconvenient to dispose the parent, it will be better